### ISSUE TITLE
Conditioning on round floats (such as 0.0 or 4.0) does not work, do we really have to condition on 0.001 or 4.0001 instead?

### ISSUE THREAD

**Issue by gcskoenig** [pyro-ppl/numpyro#1399]
For the following simple model I have problems to condition on round values such as 0.0 or 4.0.

```
import numpyro
import jax.numpy as jnp
from jax.random import PRNGKey

def model_fixed(y_obs=None, rng_key=None):
    x = 1.0
    u_y_dist = numpyro.distributions.Normal(0, 1)
    u_y = numpyro.sample('u_y', u_y_dist, rng_key=rng_key)

    y_dist = numpyro.distributions.Normal(u_y + x, scale=0.0001)
    y = numpyro.sample('y', y_dist, obs=y_obs, rng_key=rng_key)
    return x, u_y, y
```

If I sample `model_fixed(y_obs=jnp.array(0.00, dtype=jnp.float32), rng_key=PRNGKey(0))`conditioning does not work, i.e. the model samples from the normal distribution. However, if I condition on `y_obs=jnp.array(0.01, dtype=jnp.float32)` the conditioning works as expected. As a consequence inference of `u_y`works as expected for `y_obs=0.01`but not for `y_obs=0.00`.

How can I avoid this problem?

The problem is super subtle and it took me some time to figure it out. If the framework cannot handle round floats it should at least make the user aware of that.

**Comment by gcskoenig** [comment]
Fiddling around a bit more the problem rather seems to be that the conditioned upon value is probably too unrealistic for the model to handle. For instance for `1.99` the conditioning works, but not for `2.01`.

It could be helpful to print some warning to make the user aware?

**Comment by fehiepsi** [comment]
Hi @gcskoenig, could you clarify what you mean by `conditioning works/does not work`? How can I see the problem? What's the expected output if I run `model_fixed(y_obs=jnp.array(0.00, dtype=jnp.float32), rng_key=PRNGKey(0))`?

**Comment by martinjankowiak** [comment]
@gcskoenig are you running inference (e.g. HMC)? simply writing 'obs=...` doesn't do inference. it merely fixes the observed value.

**Comment by gcskoenig** [comment]
@martinjankowiak Yes, I know. I left the following MCMC code out because the error occurs before performing inference.

To make it more clear: When setting `obs=..` I expect the value of the respective variable/sample to be fixed to the specified value, which does not work here. Instead, the sample statement produces values as if no `obs=...`would be specified.

**Comment by fehiepsi** [comment]
I think it is a bug. Outside of handler stack, the sample primitive ignores the obs keywork: https://github.com/pyro-ppl/numpyro/blob/1dbe63f72f1cd8befb76d0b8e7b4f284e7852404/numpyro/primitives.py#L195 Do you want to submit a fix? (Just add an if/else statement to return obs if it is not None.)

**Comment by martinjankowiak** [comment]
yeah if you notice no where in the `numpyro` codebase will you see `rng_key` specified in a model. you probably wanted something more like this:
```
import numpyro
import jax.numpy as jnp
from jax.random import PRNGKey

def model_fixed(y_obs=None):
    x = 1.0
    u_y_dist = numpyro.distributions.Normal(0, 1)
    u_y = numpyro.sample('u_y', u_y_dist)

    y_dist = numpyro.distributions.Normal(u_y + x, scale=0.0001)
    y = numpyro.sample('y', y_dist, obs=y_obs)
    return x, u_y, y

for seed in range(3):
    with numpyro.handlers.seed(rng_seed=seed):
        print(model_fixed(1.0))
```

**Comment by gcskoenig** [comment]
@fehiepsi I can do that later today

@martinjankowiak thanks for the hint, looks much more smooth. However, I don't think that would fix the problem. Originally I did not specify the rng_seed in the model and was confused why inference would work as desired for certain values and for other's it did not. Then I added the seed to the model for debugging. I will check later whether the problem is resolved.

**Comment by gcskoenig** [comment]
I would suppose that the sample statement not taking the evidence into account is a potential cause for inference ignoring the provided evidence. 

Using `numpyro.handlers.seed()` the issue is fixed (and inference works), so there must have been another problem as well. Thank you very much for your quick help @martinjankowiak @fehiepsi!

Do you wish to support specifying the `rng_key` in the model/the bug to be fixed? Or would you like to mention that this is not supported somewhere in the documentation, e.g. [here](https://num.pyro.ai/en/stable/getting_started.html#frequently-asked-questions). 

In any case, thank you for developing `numpyro`, great stuff :)

**Comment by fehiepsi** [comment]
Under seed handler, the line `if not _PYRO_STACK:` above will not be triggered so your code will run as expected. The issue happens when you are not using any handler (including `seed`) and `obs` is not None. I think adding an `if/else` statement there would solve the topic issue.

**Comment by gcskoenig** [comment]
I created a fix:

```
    # if there are no active Messengers and no obs specified, we just draw a sample and return it as expected:
    if not _PYRO_STACK:
        if obs is None:
            return fn(rng_key=rng_key, sample_shape=sample_shape)
        else:
            return obs
```

When I try to push my bugfix branch to origin, I get the following error

```
remote: Permission to pyro-ppl/numpyro.git denied to gcskoenig.
fatal: unable to access 'https://github.com/pyro-ppl/numpyro.git/': The requested URL returned error: 403
```

I followed the standard workflow, i.e.

```
git clone [https-link]
git checkout -b bugfix-1399
git add *
git commit -m "..."
git push --set-upstream origin bugfix-1399
```

What am I doing wrong? E.g. `git remote add upstream https://github.com/pyro-ppl/numpyro` did not help. Cheers

**Comment by fehiepsi** [comment]
I think you need to fork numpyro first, add a remote pointing to your fork, push your branch to that remote, and finally make a pull request. You can't push a branch to this repository directly.
